nixops/state/json_file.py

```diff
@@ -0,0 +1,356 @@
+
+    def close(self):
+        pass
+
+    ###############################################################################################
+    ## Deployment
+
+    def query_deployments(self):
+        """Return the UUIDs of all deployments in the database."""
+
+        return self.db.read()["deployments"].keys()
+
+    def get_all_deployments(self):
+        """Return Deployment objects for every deployment in the database."""
+        uuids = self.query_deployments()
+        res = []
+        for uuid in uuids:
+            try:
+                res.append(self.open_deployment(uuid=uuid))
+            except nixops.deployment.UnknownBackend as e:
+                sys.stderr.write("skipping deployment ‘{0}’: {1}\n".format(uuid, str(e)))
+        return res
+
+    def _find_deployment(self, uuid=None):
+        all_deployments = self.db.read()["deployments"]
+
+        found = []
+
+        # if nothing exists no reason to check for things
+        if not all_deployments:
+            return None
+
+        if not uuid:
+            found = filter(lambda(id): all_deployments[id]["attributes"].get("name"), all_deployments)
+        else:
+            found = filter(lambda(id): id == uuid, all_deployments)
+            if not found:
+                found = filter(lambda(id): all_deployments[id]["attributes"].get("name") == uuid, all_deployments)
+            if not found:
+                found = filter(lambda(id): id.startswith(uuid), all_deployments)
+
+        if not found:
+            return None
+
+        if len(found) > 1:
+            if uuid:
+                raise Exception("state file contains multiple deployments with the same name, so you should specify one using its UUID")
+            else:
+                raise Exception("state file contains multiple deployments, so you should specify which one to use using ‘-d’, or set the environment variable NIXOPS_DEPLOYMENT")
+        return nixops.deployment.Deployment(self, found[0], sys.stderr)
+
+    def open_deployment(self, uuid=None):
+        """Open an existing deployment."""
+        with self.db:
+            deployment = self._find_deployment(uuid=uuid)
+            if deployment: return deployment
+            raise Exception("could not find specified deployment in state file ‘{0}’".format(self.db_file))
+
+    def create_deployment(self, uuid=None):
+        """Create a new deployment."""
+        if not uuid:
+            import uuid
+            uuid = str(uuid.uuid1())
+        with self.db:
+            state = self.db.read()
+            state["deployments"][uuid] = { "attributes": {}, "resources": {} }
+            self.db.set(state)
+        return nixops.deployment.Deployment(self, uuid, sys.stderr)
+
+    def _delete_deployment(self, deployment_uuid):
+        """NOTE: This is UNSAFE, it's guarded in nixops/deployment.py. Do not call this function except from there!"""
```

Comment:
Thanks for the heads up, but perhaps design it in such a way that it actually becomes impossible to call it from any other context.

Reply:
it wasn't written by me, but I'll think about it

Reply:
@mogorman can you comment on this?